storage: Amortize the cost of buffer appends

miretskiy opened the pull request:
While debugging slow import speeds, I noticed that runtime.growslice
as well as runtime.memmove shows up in the profiles on the
BufferingAdder.Add path.  This is due to the fact that Memfile
written to by sstable.Writer uses bytes array to which the data
was appeneded.  Instead of managing data buffer ourselves, implement
Memfile as a bufio.Buffer which nicely implements efficient
buffer allocation/re-allocation.

Release note (performance improvement): ammortize the cost of
buffer appends in Memfile.

Commit:
storage: Amortize the cost of buffer appends

While debugging slow import speeds, I noticed that runtime.growslice
as well as runtime.memmove shows up in the profiles on the
BufferingAdder.Add path.  This is due to the fact that Memfile
written to by sstable.Writer uses bytes array to which the data
was appeneded.  The buffer was never grown, resulting in
in a buffer reallocation every time anything was written to
Memfile.  Instead of managing data buffer ourselves, implement
Memfile as a bufio.Buffer which nicely implements efficient
buffer allocation/re-allocation.

The overhead observed (in my tests) wasn't that much
but it's an overhead we should not have to pay at all.

Release note (performance improvement): ammortize the cost of
buffer appends in Memfile.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44072)
<!-- Reviewable:end -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @sumeerbhola)

___
*[pkg/storage/engine/sst_writer.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44072#-LyisBMi0VyFA-sXwOnI:-LyisBMi0VyFA-sXwOnJ:b1sgfb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/819cb406f3dd152d3b3d89b1a6d0d5d930ed4f5e/pkg/storage/engine/sst_writer.go#L196)):*
> ```Go
> // Write implements the writeCloseSyncer interface.
> func (f *MemFile) Write(p []byte) (int, error) {
> 	f.data = append(f.data, p...)
> ```

The `append` builtin does handle slice growth. See https://github.com/golang/go/blob/master/src/runtime/slice.go#L66. `bytes.Buffer` seems to use somewhat different heuristics, but I'd be surprised if they make much of a difference.

On the other hand, this PR is a bit of a simplification.


<!-- Sent from Reviewable.io -->

miretskiy commented:
Interesting.. I didn't know :)

It did show up on the profiles though: 
<img width="1391" alt="Screen Shot 2020-01-15 at 5 13 57 PM" src="https://user-images.githubusercontent.com/54641364/72536449-42ef4800-3848-11ea-88ae-67907ae1313f.png">

If anything, new profile is simpler (as you say): 
<img width="1337" alt="Screen Shot 2020-01-15 at 5 59 47 PM" src="https://user-images.githubusercontent.com/54641364/72536474-500c3700-3848-11ea-9903-d51fa77574af.png">

dt commented:
Like Peter I’m surprised that it made a difference in benchmarks but the code is cleaner either way so no objections from me. Pre-alloc’ing a 1mb or something buffer might be nice too (but I still doubt anything in the sstable production path is really going g to matter all that much until the ingestion side changes dramatically).

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

I wrote the buffer append, wasn't aware about `bytes.Buffer`, my bad. Did the profiles show a performance difference? Or is it just cleaner?

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44072#-:-Lyiv6jy8N2Yzt__XjIP:b-czicgs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @sumeerbhola)

___
*[pkg/storage/engine/sst_writer.go, line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44072#-LyivUeY76ut-6sKKoin:-LyivUeZ-kXD7JkjQucQ:b-nhnx7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9a6f7f0fe0e9512e7a3d1cf85a9b67fb41a83fa/pkg/storage/engine/sst_writer.go#L197)):*
> ```Go
> // Data returns the in-memory buffer behind this MemFile.
> func (f *MemFile) Data() []byte {
> 	return f.Bytes()
> ```

`Data()` isn't used in too many places, so maybe doing an `s/sstFile.Data()/sstFile.Bytes()/` is worth doing. If it looks tedious, feel free to add it as a TODO that we can tackle later.


<!-- Sent from Reviewable.io -->

dt commented:
Ideally I'd run `BenchmarkRocksDBSstFileWriter` with this writer swapped in for the rocks one and include the benchstat in the PR / commit message if we actually think it's faster -- sometimes helps to be able to back and see where gains came from.

miretskiy commented:
Good idea:

Before: 
goos: darwin
goarch: amd64
pkg: github.com/cockroachdb/cockroach/pkg/storage/engine
BenchmarkRocksDBSstFileWriter-16    	 1000000	      1330 ns/op	 777.42 MB/s
PASS


After:
goos: darwin
goarch: amd64
pkg: github.com/cockroachdb/cockroach/pkg/storage/engine
BenchmarkRocksDBSstFileWriter-16    	 1000000	      1034 ns/op	 999.54 MB/s
PASS

Process finished with exit code 0

miretskiy commented:
Thanks @dt w/ benchstat setup.  I've modified BenchmarkRocksDBSstFileWriter to use MakeIngestionSSTWriter writer writing to MemFile.  In addition, I changed the size of the values written from 100 bytes to 2048 (random) bytes.  Ran the benchmark as follows: 
 go test -bench "BenchmarkRocksDBSstFileWriter" -run nope -count 10 -benchmem -v ./pkg/storage/engine 

The before & after comparison is:
$ benchstat before.txt after.txt
name                     old time/op    new time/op    delta
RocksDBSstFileWriter-16    1.42µs ± 8%    1.34µs ± 1%   -5.70%  (p=0.000 n=9+9)

name                     old speed      new speed      delta
RocksDBSstFileWriter-16  1.44GB/s ± 9%  1.54GB/s ± 1%   +7.00%  (p=0.000 n=10+9)

name                     old alloc/op   new alloc/op   delta
RocksDBSstFileWriter-16    1.43kB ± 0%    0.71kB ± 0%  -50.11%  (p=0.000 n=10+10)

name                     old allocs/op  new allocs/op  delta
RocksDBSstFileWriter-16      0.00           0.00          ~     (all equal)

It looks like the new code is a bit faster, but a lot better in allocations.

miretskiy commented:
bors r+

irfansharif commented on pkg/storage/engine/sst_writer.go:
```diff
@@ -178,7 +179,7 @@ func (fw *SSTWriter) Close() {
 // SSTWriter.
 type MemFile struct {
-	data []byte
+	bytes.Buffer
```
Random drive by comment: embedding it this way inherits all https://golang.org/pkg/bytes/#Buffer methods, which I don't believe is something we want.
